File: core/src/main/java/info/nightscout/androidaps/utils/ui/SingleClickButton.java
package info.nightscout.androidaps.utils.ui;

import android.app.Activity;
import android.content.Context;
import android.os.SystemClock;
import androidx.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import info.nightscout.androidaps.logging.StacktraceLoggerWrapper;

/**
 * Created by mike on 22.12.2017.
 */

public class SingleClickButton extends com.google.android.material.button.MaterialButton implements View.OnClickListener {
    final static Logger log = LoggerFactory.getLogger(SingleClickButton.class);

    Context context;
    OnClickListener listener = null;

    public SingleClickButton(Context context) {
        super(context);
        this.context = context;
        super.setOnClickListener(this);
    }

    public SingleClickButton(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.context = context;
        super.setOnClickListener(this);
    }

    public SingleClickButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        this.context = context;
        super.setOnClickListener(this);
    }

    @Override
    public void setOnClickListener(@Nullable OnClickListener l) {
        listener = l;
    }

    @Override
    public void onClick(final View v) {
        setEnabled(false);
        new Thread(() -> {
            SystemClock.sleep(1500);
            Activity activity = (Activity) context;
            if (activity != null)
                activity.runOnUiThread(() -> {
                    setEnabled(true);
                    log.debug("Button enabled");
                });
        }).start();
        if (listener != null)
            listener.onClick(v);
    }
}

File: core/src/main/java/info/nightscout/androidaps/utils/ui/SingleClickButton.kt
package info.nightscout.androidaps.utils.ui

import android.content.Context
import android.util.AttributeSet
import dagger.android.HasAndroidInjector
import info.nightscout.androidaps.core.R
import info.nightscout.shared.logging.AAPSLogger
import info.nightscout.shared.logging.LTag
import javax.inject.Inject

class SingleClickButton @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = R.style.Widget_MaterialComponents_Button) :
    com.google.android.material.button.MaterialButton(context, attrs, defStyleAttr) {

    init {
        (context.applicationContext as HasAndroidInjector).androidInjector().inject(this)
    }

    @Inject lateinit var aapsLogger: AAPSLogger

    override fun performClick(): Boolean = guardClick { super.performClick() }
    override fun callOnClick(): Boolean = guardClick { super.callOnClick() }

    private fun guardClick(block: () -> Boolean): Boolean {
        isEnabled = false
        postDelayed({ isEnabled = true; aapsLogger.debug(LTag.UI, "Button enabled") }, BUTTON_REFRACTION_PERIOD)
        return block()
    }

    companion object {

        const val BUTTON_REFRACTION_PERIOD = 3000L
    }
}